File: week_2/week_2.js
//1
function calculate(min, max){
    // 請用你的程式補完這個函式的區塊    
    let num_array =[] ;
    for(let i = min ;i < max+1 ;i++){
        num_array.push(i)
    }
    console.log( num_array.reduce((a,b) => a + b ,0)) ;
    //let reducer = (previousValue, currentValue) => previousValue + currentValue;
    //console.log(num_array.reduce(reducer));
    }
calculate(1, 3); // 你的程式要能夠計算 1+2+3，最後印出 6
calculate(4, 8); // 你的程式要能夠計算 4+5+6+7+8，最後印出 30
//*****************************************************/
//2
function avg(data){
    // 請用你的程式補完這個函式的區塊
    let sal_array =[]
    for(let i =0 ;i<data["employees"].length;i++){
        let sal = data["employees"][i]['salary']
        sal_array.push(sal)
    }
    let tot  = sal_array.reduce((a,b)=>a+b ,0)
    console.log(tot/sal_array.length)
    }
avg({
    "count":3,
    "employees":[
    {
    "name":"John",
    "salary":30000
    },
    {
    "name":"Bob",
    "salary":60000
    },
    {
    "name":"Jenny",
    "salary":50000
    }
    ]
}); // 呼叫 avg 函式

//*****************************************************/
//3
function maxProduct(nums){
    // 請用你的程式補完這個函式的區塊
    
    nums.sort(function(a, b) {
        return b - a;
      });
    let Max = nums[0]*nums[1]
    console.log(Max)
    }
maxProduct([5, 20, 2, 6]) // 得到 120
maxProduct([10, -20, 0, 3]) // 得到 30
maxProduct([-1, 2]) // 得到 -2
maxProduct([-1, 0, 2]) // 得到 0


//*****************************************************/
//4
function twoSum(nums, target){
    // your code here
    let temp_array =[]
    for(let i = 0 ;i<nums.length ;i++){
        for(let j = 0 ;j<nums.length ;j++){
            if(nums[i]+nums[j] == target){
                temp_array.push(i)}
    }}
    return temp_array
    //console.log(temp_array)
    }
let result=twoSum([2, 11, 7, 15], 9);
console.log(result); // show [0, 2] because nums[0]+nums[2] is 9
// 個人資測
// let result2=twoSum([5, 20, 1, 2], 3);
// console.log(result2); // show [2, 3] because nums[2]+nums[3] is 3

//*****************************************************/
//5

function maxZeros(nums){
    // 請用你的程式補完這個函式的區塊
    let temp = 0
    let big = 0
    for(let i =0 ;i<nums.length ;i++){
        if(nums[i]==0){
            temp++
            if(big<temp){
                big = temp
            }
        }
        else{temp = 0
        }
    }
    console.log(big)
    }
maxZeros([0, 1, 0, 0]); // 得到 2
maxZeros([1, 0, 0, 0, 0, 1, 0, 1, 0, 0]); // 得到 4
maxZeros([1, 1, 1, 1, 1]); // 得到 0
maxZeros([0, 0, 0, 1, 1]) // 得到 3

File: week_2/week_2.py
# -*- coding: utf-8 -*-
"""
Created on Mon Jan  3 09:52:30 2022

@author: nash1
"""

#1
def calculate(min, max):
    num_array = []
    for i in range(min,max+1):
        num = i
        num_array.append(num)
    print(sum(num_array))
# 請用你的程式補完這個函式的區塊
calculate(1, 3) # 你的程式要能夠計算 1+2+3，最後印出 6
calculate(4, 8) # 你的程式要能夠計算 4+5+6+7+8，最後印出 30


#2
def avg(data):
    # 請用你的程式補完這個函式的區塊
    sal_array = []
    for i in range(len(data["employees"])):
        #print(data["employees"][i]['salary'])
        sal = data["employees"][i]['salary']
        sal_array.append(sal)
    print(sum(sal_array)/len(data["employees"]))

avg({
"count":3,
"employees":[
{
"name":"John",
"salary":30000
},
{
"name":"Bob",
"salary":60000
},
{
"name":"Jenny",
"salary":50000
}
]
}) # 呼叫 avg 函式


#3
def maxProduct(nums):
# 請用你的程式補完這個函式的區塊
    new = sorted(nums ,reverse = True)
    big = new[0]*new[1]
    print(big)
    
maxProduct([5, 20, 2, 6]) # 得到 120
maxProduct([10, -20, 0, 3]) # 得到 30
maxProduct([-1, 2]) # 得到 -2
maxProduct([-1, 0, 2]) # 得到 0

#4.
def twoSum(nums, target):
    #假設兩兩相加且元素位置不重複算法
    temp_array =[]
    # your code here
    for i in range(len(nums)):
        for j in range(len(nums)):
             if nums[i]+nums[j] == target:
                 temp = i
                 temp_array.append(temp)
    return temp_array
                    
result=twoSum([2, 11, 7, 15], 9)
print(result) # show [0, 2] because nums[0]+nums[2] is 9

result=twoSum([5, 20, 1, 2], 3)
print(result) # show [2, 3] because nums[2]+nums[3] is 3


def maxZeros(nums):
# 請用你的程式補完這個函式的區塊
    temp = 0
    big =0
    for i in range(len(nums)):
        if nums[i] == 0:
            temp +=1
            if big<temp:
                big = temp
        else:
            temp = 0
    print(big)
    
maxZeros([0, 1, 0, 0]) # 得到 2
maxZeros([1, 0, 0, 0, 0, 1, 0, 1, 0, 0]) # 得到 4
maxZeros([1, 1, 1, 1, 1]) # 得到 0
maxZeros([0, 0, 0, 1, 1]) # 得到 3










